ENR

jannikluhn opened the pull request:
Closes #583. Still WIP: A couple of additional tests are needed and we have to use compressed public keys (working on https://github.com/ethereum/eth-keys/issues/55 now).

### How was it fixed?

- There's an ENR class that exposes the ENR key value pairs via a dict interface, along the sequence number and signature. If the signature is `None` it is considered unsigned. A signed copy can be created with `to_signed_enr`.
- There are special sedes classes `ENRSedes` and `ENRContentSedes` to serialize/deserialize an ENR with or without the signature. Values are (de)serialized as binaries, unless there's something better in `ENR_KEY_SEDES_MAPPING`. 
- Identity schemes (responsible for signing and validating signatures) are represented by children of `ENRIdentityScheme` which define a method to check if they can be used for a given ENR or not (usually by checking if the ENR contains a public key in the right format). The ENR will use the first matching identity scheme it can find (there will probably only be a few).

#### Cute Animal Picture

![put a cute animal picture link inside the parentheses](https://user-images.githubusercontent.com/29854669/57213104-17903a00-6fe6-11e9-87dc-c1626eff5754.jpg)

jannikluhn commented:
Updated to use the new eth-keys API introduced in https://github.com/ethereum/eth-keys/pull/56 and https://github.com/ethereum/eth-keys/pull/58. I still have to add the new version in `setup.py`, but I can only do this once the new changes have been released.

I also restructured the code a little in order to have two separate classes for signed and unsigned ENRs (instead of an `is_signed` property and a `signature` that may be `None`).

hwwhww commented on p2p/discv5/identity_schemes.py:
```diff
@@ -0,0 +1,91 @@
+
+    @classmethod
+    @abstractmethod
+    def create_signature(cls, enr: "BaseENR", private_key: Any) -> bytes:
```
```suggestion
    def create_signature(cls, enr: "BaseENR", private_key: PrivateKey) -> bytes:
```

jannikluhn replied:
I'd actually like to make it use `bytes` instead. The reason is that different identity schemes will use different private key formats and `bytes` would be the lowest most common denominator.

hwwhww commented on p2p/discv5/enr.py:
```diff
@@ -0,0 +1,236 @@
+
+class UnsignedENR(BaseENR, ENRContentSedes):
+
+    def to_signed_enr(self, private_key: Any) -> "ENR":
```
```suggestion
    def to_signed_enr(self, private_key: PrivateKey) -> "ENR":
```

hwwhww commented on p2p/discv5/identity_schemes.py:
```diff
@@ -0,0 +1,91 @@
+from eth_keys.datatypes import (
+    PrivateKey,
+    PublicKey,
+    NonRecoverableSignature,
```
ImportError: cannot import name 'NonRecoverableSignature' from 'eth_keys.datatypes' (/home/circleci/repo/.tox/py37-p2p/lib/python3.7/site-packages/eth_keys/datatypes.py)

pipermerriam commented on p2p/discv5/enr.py:
```diff
@@ -0,0 +1,236 @@
+        serialized_sequence_number = big_endian_int.serialize(enr.sequence_number)
+        serialized_keys = tuple(binary.serialize(key) for key in enr.keys())
+        value_serializers = tuple(
+            ENR_KEY_SEDES_MAPPING.get(key, binary)
```
This probably deserves a comment.  Not intuitive to me (though I'm not familiar to the protocol) why we'd fall back to the `binary` sedes.  Maybe something explaining the basic structure of the record and why we do this fallback?

jannikluhn replied:
It doesn't really matter as we don't use the value in any way, we could even use `big_endian_int`. `binary` seemed to make the most sense as it doesn't try to interpret, it's basically an identity function. I put it in a constant and added a comment.

pipermerriam commented on p2p/discv5/enr.py:
```diff
@@ -0,0 +1,236 @@
+    def serialize(cls, enr: "ENR") -> Tuple[bytes, ...]:
+        serialized_signature = binary.serialize(enr.signature)
+        serialized_content = ENRContentSedes.serialize(enr)
+        return (serialized_signature,) + serialized_content
```
Not necessarily better but I found [`toolz.cons`](https://toolz.readthedocs.io/en/latest/api.html#toolz.itertoolz.cons) recently which does exactly this.  I believe it is lazy meaning that it has to be combined with an outer `tuple` cast in this scenario which isn't exactly ideal.

pipermerriam commented on p2p/discv5/enr.py:
```diff
@@ -0,0 +1,236 @@
+        if not applicable_identity_schemes:
+            raise ValidationError("No known identity scheme used")
+
+        return applicable_identity_schemes[0]
```
Is it intentional that this just blindly takes the first one?  If so I think that deserves a comment indicating that we're aware that this discards the extra ones and that is an intentional design decision.

Second, this function could be re-structured as:

```python
for scheme in identity_schemes:
    if scheme.is_applicable(self):
        return scheme
else:
    raise ValidationError(...)
```

This should be slightly more performant as well.

jannikluhn replied:
Yes, the idea is that `identity_schemes` is ordered by preference, so we just pick the one that we like the most. I don't expect there to be many, probably not more than two.

I changed the name to `get_preferred_identity_scheme` and added a comment to the definition of `identity_schemes` about the order. I hope that makes it more clear.

jannikluhn replied:
Spec has just been updated to explicify the used/supported identity scheme: https://github.com/ethereum/EIPs/commit/9b1c22d2850cc8efb7aabb2aba4098873ed0d5a9

As the change doesn't appear to be final yet, I'm gonna wait a little before updating, but definitely do it before merging.

pipermerriam commented on p2p/discv5/enr.py:
```diff
@@ -0,0 +1,236 @@
+    def __eq__(self, other: Any) -> bool:
+        return (
+            other.__class is self.__class__ and
+            dict(other) == dict(self) and
```
Since it appears that ENR records are immutable mappings, what do you think about implementing `__hash__` on the class and implementing this as `type(other) is type(self) and hash(other) == hash(self)` and capturing all of the value equality in the `__hash__` implementation.

carver replied:
Hash is not guaranteed to be unique so that equality test gets you some false positives (and prevents subclasses from matching, which means false negatives, I suppose). Implementing hash is still worthwhile performance-wise, since most equality checks that matter will check the hash first anyways. Though I don't think it is strictly necessary for this pr.

jannikluhn replied:
Added a hash function, but stayed away from using it in `__eq__` as false positives sound dangerous.

pipermerriam commented on p2p/discv5/enr.py:
```diff
@@ -0,0 +1,236 @@
+                 sequence_number: int,
+                 kv_pairs: Mapping[bytes, Any]) -> None:
+        self._sequence_number = sequence_number
+        self._kv_pairs = kv_pairs
```
Since it appears the dict API implemented on the ENR is intended to make it immutable, maybe this value should be isolated by casting it to `dict` to ensure any ties with the outside object dont't result in mutation of the ENR values.

pipermerriam commented on p2p/discv5/identity_schemes.py:
```diff
@@ -0,0 +1,91 @@
+            is_valid = False
+
+        if not is_valid:
+            raise ValidationError("Invalid signature")
```
Seems like you could just raise this from within the `except` block above to avoid the need to store an `is_valid` variable.

jannikluhn replied:
`verify_signature` can either return `False` or raise a `BadSignature`, depending on the way the signature is invalid. So there isn't away to avoid the variable without copying the `raise ValidationError` line which I'd like to avoid.

jannikluhn replied:
> That suggests the verify_signature api should be changed if possible to either convert the exception case to return false. I'm on mobile so haven't really done my homework on feasibility or why the api was designed that way.

Looks like what I said is not accurate, at least not after the changes in https://github.com/ethereum/eth-keys/pull/58 (as we don't try to recover the signer anymore). So `verify_msg` should never fail as far as I can see. However, the signature can still be broken, but this would raise an exception when creating the signature. I updated the function accordingly to cover the two failure cases. It still needs the `is_valid` variable, but it looks less weird now.

pipermerriam commented on tests/p2p/discv5/test_enr.py:
```diff
@@ -0,0 +1,138 @@
+)
+
+
+OFFICIAL_ENR_RLP = decode_hex(
```
Can you add a comment linking to the source of this test vector?

jannikluhn commented:
Opened a meta issue here (I guess I should have started with that): https://github.com/ethereum/trinity/issues/668

jannikluhn commented on p2p/discv5/enr.py:
```diff
@@ -0,0 +1,236 @@
+ENR_KEY_SEDES_MAPPING = {
+    b"id": binary,
+    b"secp256k1": Binary(min_length=33, max_length=33),
+    b"ip": binary,
```
Did it the proper way and added a special sedes object for IP addresses.

jannikluhn commented:
I've addressed the feedback and now I'm waiting for ethereum/eth-keys#58 and a subsequent eth-keys release. Then I'll have a look at the outstanding CI errors.

pipermerriam commented:
That suggests the verify_signature api should be changed if possible to either convert the exception case to return false. I'm on mobile so haven't really done my homework on feasibility or why the api was designed that way.

Sent from ProtonMail mobile

-------- Original Message --------
On May 26, 2019, 4:07 PM, jannikluhn wrote:

> @jannikluhn commented on this pull request.
>
> ---------------------------------------------------------------
>
> In [p2p/discv5/identity_schemes.py](https://github.com/ethereum/trinity/pull/586#discussion_r287599483):
>
>> +        signature = private_key.sign_msg_non_recoverable(message)
> +        return bytes(signature)
> +
> +    @classmethod
> +    def validate_signature(cls, enr: "ENR") -> None:
> +        public_key = PublicKey.from_compressed_bytes(enr[cls.public_key_enr_key])
> +        signature = NonRecoverableSignature(enr.signature)
> +        message = enr.get_signing_message()
> +
> +        try:
> +            is_valid = signature.verify_msg(message, public_key)
> +        except BadSignature:
> +            is_valid = False
> +
> +        if not is_valid:
> +            raise ValidationError("Invalid signature")
>
> verify_signature can either return False or raise a BadSignature, depending on the way the signature is invalid. So there isn't away to avoid the variable without copying the raise ValidationError line which I'd like to avoid.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, [view it on GitHub](https://github.com/ethereum/trinity/pull/586?email_source=notifications&email_token=AAGJHAVZOCEIGL3VUZMLFCDPXKKR3A5CNFSM4HK56BGKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBZWUFBA#discussion_r287599483), or [mute the thread](https://github.com/notifications/unsubscribe-auth/AAGJHARB5O6C3IK65PIMSULPXKKR3ANCNFSM4HK56BGA).

fjl commented:
Very nice to see Trinity will have ENR support :).
Note that EIP-778 was [changed one more time](https://github.com/ethereum/EIPs/commit/5a0665ca9cece6172087fe9eadcfd1c6d22ba9c9) to add "tcp6", "udp6" keys. They allow
relaying IPv6-specific port numbers. 

We have also decided to use base64url encoding (without padding) instead of standard base64.
Base64 without padding is not supported in the Python standard library, but you can use these
functions to do it: https://github.com/fjl/p2p-drafts/commit/23ed3427007e19a3e6f81202e3e967ddd74ef9e7#diff-09ba11ec95dfcee8269304dd2cc0abf3R156

jannikluhn commented:
- Updated to latest version of ENR spec (thanks @fjl)
- reworked tests: clear separation between ENR and identity scheme
- cleaned up commit history

Unfortunately, now everything breaks due to a version conflict. Pretty much all `eth-*` libraries require `eth_keys < 3.0.0`, but here we need `eth_keys >= 3.0.0`. I guess there's no other way but going through all of them and make a new release? `3.0.0` *should* be backwards compatible.

carver commented:
> Unfortunately, now everything breaks due to a version conflict. Pretty much all `eth-*` libraries require `eth_keys < 3.0.0`, but here we need `eth_keys >= 3.0.0`. I guess there's no other way but going through all of them and make a new release? `3.0.0` _should_ be backwards compatible.

Yeah, sorry, this is fixed in #699 and was released so it doesn't affect fresh installs. But it hasn't been merged in yet. I'll ping you when it gets merged in, and the rebase should be painless.

carver commented:
@jannikluhn Okay, a rebase should fix the eth-keys issue now.

jannikluhn commented:
@carver that didn't do the trick. The problem is that this PR needs `eth-keys >= 3.0.0`, everything else `eth-keys < 3.0.0`. So I think there's no other way than updating all of them to use `< 4.0.0`.

carver commented:
Ah, right. @pipermerriam instead of propagating a chain of eth-keys releases up to trinity (I think at least these would need new releases: eth-account, web3, py-evm), maybe we can just do a patch release that includes everything that's in 0.3.0?

pipermerriam commented:
`v0.2.3` is now available and equal to `v0.3.0`

carver commented on setup.py:
```diff
@@ -14,6 +14,7 @@
         # cryptography does not use semver and allows breaking changes within `0.3` version bumps.
         "cryptography>=2.5,<2.7",
         "eth-hash>=0.1.4,<1",
+        "eth-keys>=0.3.0,<0.4.0",
```
@jannikluhn Can you try again with

```suggestion
        "eth-keys>=0.2.3,<0.3.0",
```

carver commented:
Hm, interesting, I think this is a legit issue. The new `eth-keys` is treating a valid signature as invalid?

```
_ test_rpc_against_fixtures[/home/circleci/repo/fixtures/BlockchainTests/bcValidBlockTest/SimpleTx3.json:SimpleTx3_Frontier:Frontier] _
[gw1] linux -- Python 3.6.8 /home/circleci/repo/.tox/py36-rpc-blockchain/bin/python

event_bus = <trinity.endpoint.TrinityEventBusEndpoint object at 0x7f2f8c0a39b0>
chain_fixture = {'_info': {'comment': '', 'filledwith': 'testeth 1.5.0.dev2-68+commit.4a43125d.dirty', 'lllcversion': 'Version: 0.5.0-...80832fefd8808454c98c8142a056e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421880102030405060708c0c0', ...}
fixture_data = ('/home/circleci/repo/fixtures/BlockchainTests/bcValidBlockTest/SimpleTx3.json', 'SimpleTx3_Frontier', 'Frontier')

    @pytest.mark.asyncio
    async def test_rpc_against_fixtures(event_bus, chain_fixture, fixture_data):
        rpc = RPCServer(
            initialize_eth1_modules(MainnetFullChain(None), event_bus)
        )
    
        setup_result, setup_error = await call_rpc(rpc, 'evm_resetToGenesisFixture', [chain_fixture])
        # We need to advance the event loop for modules to be able to pickup the new chain
        await asyncio.sleep(0)
        assert setup_error is None and setup_result is True, "cannot load chain for {0}".format(fixture_data)  # noqa: E501
    
        await validate_accounts(rpc, chain_fixture['pre'])
    
        for block_fixture in chain_fixture['blocks']:
            should_be_good_block = 'blockHeader' in block_fixture
    
            if 'rlp_error' in block_fixture:
                assert not should_be_good_block
                continue
    
            block_result, block_error = await call_rpc(rpc, 'evm_applyBlockFixture', [block_fixture])
    
            if should_be_good_block:
>               assert block_error is None
E               AssertionError: assert 'Invalid Signature' is None

block_error = 'Invalid Signature'
block_fixture = {'blockHeader': {'bloom': '0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000...ce': '0x00', ...}, {'data': '0x', 'gasLimit': '0x5208', 'gasPrice': '0x01', 'nonce': '0x03', ...}], 'uncleHeaders': []}
block_result = None
chain_fixture = {'_info': {'comment': '', 'filledwith': 'testeth 1.5.0.dev2-68+commit.4a43125d.dirty', 'lllcversion': 'Version: 0.5.0-...80832fefd8808454c98c8142a056e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421880102030405060708c0c0', ...}
event_bus  = <trinity.endpoint.TrinityEventBusEndpoint object at 0x7f2f8c0a39b0>
fixture_data = ('/home/circleci/repo/fixtures/BlockchainTests/bcValidBlockTest/SimpleTx3.json', 'SimpleTx3_Frontier', 'Frontier')
rpc        = <trinity.rpc.main.RPCServer object at 0x7f2f68721ba8>
setup_error = None
setup_result = True
should_be_good_block = True
```

jannikluhn commented:
yeah, that's my fault, this PR should fix it: https://github.com/ethereum/eth-keys/pull/62

jannikluhn commented:
Finally everything is green :tada:

I think I've addressed all comments so far, but since the last review I've restructured and extended the tests a bit. Anyone wants to have another look? Otherwise this is ready from my side.

pipermerriam commented on p2p/discv5/enr.py:
```diff
@@ -0,0 +1,336 @@
+    def serialize(cls, enr: "BaseENR") -> Tuple[bytes, ...]:
+        serialized_sequence_number = big_endian_int.serialize(enr.sequence_number)
+
+        sorted_key_value_pairs = sorted(enr.items(), key=lambda key_value: key_value[0])
```
I believe historically we've measured `operator.itemgetter(0)` to be faster and slightly more preferred.

pipermerriam commented on p2p/discv5/enr.py:
```diff
@@ -0,0 +1,336 @@
+                                 serialized_enr: Sequence[bytes]) -> None:
+        duplicates = set(
+            key for index, key in enumerate(keys)
+            if key in keys[index + 1:]
```
I think this has some unfortunate time complexity.  An `itertools.Counter` does a pretty nice job of both detecting duplicates and allowing easy reporting of which ones are duplicate.

```python
duplicates = {key for key, num in itertools.Counter(keys).items() if num > 1}
```

That one-liner does the trick pretty concisely.

jannikluhn replied:
neat! changed it

pipermerriam commented on p2p/discv5/enr.py:
```diff
@@ -0,0 +1,336 @@
+            )
+
+
+class BaseENR(Mapping[bytes, Any], ABC):
```
Are the value types truly `Any` or can that be constrained more?

jannikluhn replied:
For now it's just `int` and `bytes`, but in theory it can be anything -- we might want to add special types for, e.g., libp2p multiaddrs or public key objects in the future. So to keep it easily extendable I'd prefer to keep it `Any`.

pipermerriam commented on p2p/discv5/enr.py:
```diff
@@ -0,0 +1,336 @@
+
+    def get_identity_scheme(self) -> Type[IdentityScheme]:
+        try:
+            identity_scheme_id = self[b"id"]
```
Maybe instead of having to do this `try/except` checking at runtime it should happen during initialization and be memoized somewhere private since it's probably very natural to do that at whatever place the `b'id'` key is validated to be present.

pipermerriam commented on p2p/discv5/identity_schemes.py:
```diff
@@ -0,0 +1,104 @@
+    )
+
+
+identity_scheme_registry: Dict[bytes, Type["IdentityScheme"]] = {}
```
We probably still use this pattern in a number of places but I've grown to think that we should be using a registry pattern closer to something like this.

```
class Registry(collections.UserDict):
    def register_thing(self, thing) -> None:
        ...  # business logic here
        self[thing_key] = thing

default_registry = Registry()
register_thing = default_registry.register_thing
```

Note that `collections.UserDict` gives us an easy mechanism for simple registries like this.

https://docs.python.org/3/library/collections.html#collections.UserDict

Note that this requires ALL apis that make use of the registry to have some mechanism for passing in a custom registry.  This pattern should result in easier isolation during testing as well as more flexibility when a library needs to override the base registry.  I also think this approach has minimal extra overhead with respect to implementation and user facing API since the actualy underlying `Registry` API can be maintained as private.

jannikluhn replied:
Sounds reasonable. I've implemented something like this, but I'm not sure if I fully understood. Please have a look if that's how you imagined it.

pipermerriam commented on tests/p2p/discv5/test_enr.py:
```diff
@@ -0,0 +1,312 @@
+            b"value1",
+            b"id",
+        ]),
+    )
```
This should be parametrized using `pytest.mark.parametrize` so it runs as 4 different tests.

pipermerriam commented on p2p/discv5/identity_schemes.py:
```diff
@@ -0,0 +1,115 @@
+        pass
+
+
+@default_identity_scheme_registry.register
```
You might expose the `register` function as a top level module item that users can import to register a new identity scheme but that's only relevant if you want to expose this as a public API.  For now I think it's fine to keep it private.

pipermerriam commented on tests/p2p/discv5/test_enr.py:
```diff
@@ -0,0 +1,345 @@
+
+
+@pytest.fixture
+def test_identity_scheme_registry(mock_identity_scheme):
```
because pytest is *greedy* I've tended to avoid naming anything in a module that pytest scrapes with `test_` (or classes that start with `Test`) because it sometimes tries to run them as tests.  Naming schemes to get around this that seem appropriate are just to call this `identity_scheme_registry` or `identity_scheme_registry_for_testing`
